File: leetcode/1406-stone-game-iii/solution.py
from typing import List


class Solution:
    def stoneGameIII(self, stoneValue: List[int]) -> str:
        n = len(stoneValue)
        dp = [0] * (n + 1)

        for i in range(n - 1, -1, -1):
            best = float('-inf')
            take = 0
            for x in range(1, 4):
                if i + x > n:
                    break
                take += stoneValue[i + x - 1]
                best = max(best, take - dp[i + x])
            dp[i] = best

        if dp[0] > 0:
            return "Alice"
        elif dp[0] < 0:
            return "Bob"
        return "Tie"

File: leetcode/1406-stone-game-iii/solution.cpp
#include <string>
#include <vector>
#include <climits>
#include <algorithm>
using namespace std;

class Solution {
public:
    string stoneGameIII(vector<int>& stoneValue) {
        int n = stoneValue.size();
        vector<int> dp(n + 1, 0);

        for (int i = n - 1; i >= 0; i--) {
            int best = INT_MIN;
            int take = 0;
            for (int x = 1; x <= 3 && i + x <= n; x++) {
                take += stoneValue[i + x - 1];
                best = max(best, take - dp[i + x]);
            }
            dp[i] = best;
        }

        if (dp[0] > 0) return "Alice";
        if (dp[0] < 0) return "Bob";
        return "Tie";
    }
};
